File: python/756. Pyramid Transition Matrix.py
class Solution(object):
    def pyramidTransition(self, bottom, allowed):
        """
        :type bottom: str
        :type allowed: List[str]
        :rtype: bool
        """
        allowed = set(allowed)
        def dfs(bottom, nextStr):
            if len(bottom)==1:
                return True
            if len(bottom) - len(nextStr)>1:
                s = bottom[len(nextStr): len(nextStr)+2]
                for i in range(7):
                    ch = chr(ord('A') + i)
                    if s+ch in allowed and dfs(bottom, nextStr+ch):
                        return True
                return False
            else:
                return dfs(nextStr, "")
        
        return dfs(bottom, "")

if __name__ == '__main__':
    bottom = 'AABA'
    allowed = ["AAA", "AAB", "ABA", "ABB", "BAC"]
    su = Solution()
    print su.pyramidTransition(bottom, allowed)        

File: c++/756. Pyramid Transition Matrix.cpp
class Solution {
public:
    bool pyramidTransition(string bottom, vector<string>& allowed) {
        unordered_set<string> st(allowed.begin(), allowed.end());
        return dfs(bottom, "", st);
    }

    bool dfs(string bottom, string nextStr, unordered_set<string>& allowed) {
        if (bottom.size() == 1) return true;

        if (bottom.size() - nextStr.size() > 1) {
            string str = bottom.substr(nextStr.size(), 2);
            for (char c = 'A'; c < 'H'; c++) {
                auto it = allowed.find(str + c);
                if (it!=allowed.end() && dfs(bottom, nextStr + c, allowed)) {
                    return true;
                }
            }
            return false;
        } else {
            return dfs(nextStr, "", allowed);
        }
    }

};

// 复杂一点的dfs， 通过判断上一层字符串的长度来做搜索分支